src/raw_rtde/connection.rs

```diff
@@ -0,0 +1,891 @@
+use std::io::{Cursor, ErrorKind, Read, Write};
+use std::mem::{discriminant, Discriminant};
+use std::net::SocketAddr;
+use std::ops::Add;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+use byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
+use mio::net::TcpStream;
+use mio::{event, Events, Interest, Poll, Token, Waker};
+
+use crate::raw_rtde::packet_data::{ControllerVersion, IoSetup, RTDEData};
+
+const RTDE_SEND_TIMEOUT: Duration = Duration::from_millis(50);
+const RTDE_WAIT_FOR_REPLY_TIMEOUT: Duration = Duration::from_millis(2000);
+
+enum RTDECommand {
+    ConfigureInputs = b'I' as isize,
+    ConfigureOutputs = b'O' as isize,
+    Start = b'S' as isize,
+    Pause = b'P' as isize,
+    RtdeDataPacket = b'U' as isize,
+    RequestProtocolVersion = b'V' as isize,
+    GetControllerVersion = b'v' as isize,
+}
+const WOKEN_BEFORE_READ_PSEUDO_CMD: u8 = 255;
+
+/// A low-level class to control RTDE at the level of single packets.
+///
+/// Except for tests it's recommended to use `RawRTDE` to send RTDE over TCP/IP,
+/// as the UR robot expects to receive it.
+pub struct GenericRawRTDE<T: Read + Write + event::Source> {
```

Comment:
Drop the bounds from type definitions, they are unnecessary and only cause annoyance. Use bounds only when actually used (i.e., in impls).

Reply:
This works fine for GenericRawRTDE, but dropping the bounds in RTDEThread leads to:
```
error[E0367]: `Drop` impl requires `T: connection::RawRTDETrait` but the struct it is implemented for does not
   --> ur_rtde/src/rtde_thread/rtde_thread.rs:226:8
    |
226 |     T: RawRTDETrait + Send,
    |        ^^^^^^^^^^^^
    |
note: the implementor must specify the same requirement
   --> ur_rtde/src/rtde_thread/rtde_thread.rs:27:1
    |
27  | / pub struct RTDEThread<T>
28  | | {
29  | |     shared_state: Arc<Mutex<SharedState>>,
30  | |     condition: Arc<Condvar>,
...   |
34  | |     phantom: PhantomData<T>, // In case we need T in the future.
35  | | }
    | |_^
```
(and a similar error for `Send` trait)

Is it that if I need the `Drop` trait, I need to keep the bounds in the type definition?

Reply:
Is there a reason for the bound in the `Drop` impl? If it's actually used, then you'll need to keep it in the declaration as well. If it isn't, then remove it from both places.